kratos/input_output/base_64_encoded_output.h

```diff
@@ -0,0 +1,306 @@
+
+            for (IndexType i = 0; i < number_of_triplets; ++i) {
+                EncodeTriplet(mrOStream, {*itr_byte_triplet++, *itr_byte_triplet++, *itr_byte_triplet++}, 0);
+            }
+
+            number_of_written_bytes += number_of_triplets * 3;
+
+            // now fill the mByteTriplet with the remaining bytes of the data
+            const IndexType remaining_bytes = raw_bytes - number_of_written_bytes;
+            for (; mByteTripletIndex < remaining_bytes; ++mByteTripletIndex) {
+                mByteTriplet[mByteTripletIndex] = *itr_byte_triplet++;
+            }
+        }
+    }
+
+    ///@}
+
+private:
+    ///@name Private member variables
+    ///@{
+
+    /**
+    * @brief The output stream for encoding operations.
+    * @details This member variable represents the output stream where the encoded data will be written.
+    * It is used by the encoding functions in this class to write the encoded data.
+    */
+    std::ostream& mrOStream;
+
+    /**
+    * @brief The index of the current byte triplet being processed.
+    * @details This member variable keeps track of the current byte triplet being processed during encoding.
+    * It is used to determine the position of the bytes in the triplet and to handle padding correctly.
+    */
+    IndexType mByteTripletIndex = 0;
+
+    /**
+    * @brief The byte triplet buffer for encoding operations.
+    * @details This member variable is an array of three characters representing the current byte triplet being encoded.
+    * It is used to store the bytes before they are processed and encoded into base64 format.
+    */
+    std::array<char, 3> mByteTriplet;
+
+    /**
+    * @brief The base64 encoding character mapping.
+    * @details This static constexpr member variable is a character array that maps the values used for base64 encoding.
+    * It contains the characters 'A' to 'Z', 'a' to 'z', '0' to '9', and the '+' and '/' characters.
+    * It is used by the encoding functions to convert the binary data into base64-encoded characters.
+    */
+    constexpr static char base64Map[] = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
+
+    ///@}
+    ///@name Private static operations
+    ///@{
+
+    /**
+    * @brief Encodes a triplet of bytes into base64 format and writes it to the specified output stream.
+    * @details This function takes three bytes and encodes them using the base64 encoding scheme.
+    * The resulting encoded triplet is written to the provided output stream.
+    * @param rOutput The output stream where the encoded triplet will be written.
+    * @param rBytes An array of three bytes to encode.
+    * @param Padding The number of padding characters to append to the encoded triplet.
+    *                It should be a value between 0 and 2 (inclusive).
+    * @note The output stream should be in a valid state and open for writing.
+    */
+    static inline void EncodeTriplet(
```

Comment:
Arent static and inline conflicting?

Reply:
I don't think they are, but `inline` doesn't do anything if the function is defined inside the class definition.

Reply:
Just to be sure, removing